core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java

```diff
@@ -29,120 +31,95 @@ public class UnavailableSpaceBuilder {
-                        preBlockLength
+                // Generate current block occupancy
+                unavailableSpace.put(blockId, new OccupancySegment(
+                        timeRange.lowerEndpoint(), timeRange.upperEndpoint(), 0, blockLength)
                 );
-                unavailableSpace.put(predecessorRoute, previousBlock);
-            }
 
-            //Generating successorRoute occupancy
-            var successorRoutes = getNextRoutes(rawInfra, route);
-            for (var successorRoute : successorRoutes) {
-                var successorBlockLength = rawInfra.getRouteLength(successorRoute);
-                addRouteOccupancy(
-                        unavailableSpace,
-                        rawInfra,
-                        blockInfra,
-                        successorRoute,
-                        timeStart,
-                        timeEnd,
-                        0,
-                        successorBlockLength
-                );
-            }
+                // Generate the warnings in blocks before the ones used by other trains
+                var predecessorBlocks = getPreviousBlocks(infra, blockInfra, blockId);
+                for (var predecessorBlock : predecessorBlocks) {
+                    var preBlockLength = blockInfra.getBlockLength(predecessorBlock);
+                    unavailableSpace.put(predecessorBlock, new OccupancySegment(
+                            timeRange.lowerEndpoint(), timeRange.upperEndpoint(), 0, preBlockLength
+                    ));
 
-            //Generating rollingStock length occupancy
-            for (var successorRoute : successorRoutes) {
-                var secondSuccessorRoutes = getNextRoutes(rawInfra, successorRoute);
-                for (var secondSuccessorRoute : secondSuccessorRoutes) {
-                    var end = Math.min(
-                            Distance.fromMeters(rollingStock.getLength()),
-                            rawInfra.getRouteLength(secondSuccessorRoute)
-                    );
-                    addRouteOccupancy(unavailableSpace, rawInfra, blockInfra, secondSuccessorRoute,
-                            timeStart, timeEnd, 0, end);
+                    // Generate the sight distance requirements in the blocks before that
+                    for (var secondPredecessorBlock : getPreviousBlocks(infra, blockInfra, predecessorBlock)) {
+                        var secPreBlockLength = blockInfra.getBlockLength(secondPredecessorBlock);
+                        unavailableSpace.put(secondPredecessorBlock, new OccupancySegment(
+                                timeRange.lowerEndpoint(), timeRange.upperEndpoint(),
+                                Math.max(0, secPreBlockLength - SIGHT_DISTANCE), preBlockLength
+                        ));
+                    }
+                }
+
+                // Generate train length occupancy
+                var successorBlocks = getNextBlocks(infra, blockInfra, blockId);
```

Comment:
What if the train's length is > to the next block's length ?  Shouldn't the block after that be marked as used, for `length = min(nextNextBlockLength, Distance.fromMeters(rollingStock.getLength()) - nextBlockLength)` ?

Reply:
True, it should be. But I don't know if it's worth the added code complexity, this code is meant to be replaced very soon. It's an uncommon case and doesn't cause significant issues.